concurrency/RecoveryState.java

```diff
@@ -0,0 +1,961 @@
+            }
+            return reused;
+        }
+
+        public synchronized long reusedBytes() {
+            long reused = 0;
+            for (File file : fileDetails.values()) {
+                if (file.reused()) {
+                    reused += file.length();
+                }
+            }
+            return reused;
+        }
+
+        public synchronized void updateVersion(long version) {
+            this.version = version;
+        }
+
+        @Override
+        public synchronized XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            // stream size first, as it matters more and the files section can be long
+            builder.startObject(Fields.SIZE);
+            builder.humanReadableField(Fields.TOTAL_IN_BYTES, Fields.TOTAL, new ByteSizeValue(totalBytes()));
+            builder.humanReadableField(Fields.REUSED_IN_BYTES, Fields.REUSED, new ByteSizeValue(reusedBytes()));
+            builder.humanReadableField(Fields.RECOVERED_IN_BYTES, Fields.RECOVERED, new ByteSizeValue(recoveredBytes()));
+            builder.field(Fields.PERCENT, String.format(Locale.ROOT, "%1.1f%%", recoveredBytesPercent()));
+            builder.endObject();
+
+            builder.startObject(Fields.FILES);
+            builder.field(Fields.TOTAL, totalFileCount());
+            builder.field(Fields.REUSED, reusedFileCount());
+            builder.field(Fields.RECOVERED, recoveredFileCount());
+            builder.field(Fields.PERCENT, String.format(Locale.ROOT, "%1.1f%%", recoveredFilesPercent()));
+            if (params.paramAsBoolean("detailed", false)) {
+                builder.startArray(Fields.DETAILS);
+                for (File file : fileDetails.values()) {
+                    file.toXContent(builder, params);
+                }
+                builder.endArray();
+            }
+            builder.endObject();
+            builder.humanReadableField(Fields.TOTAL_TIME_IN_MILLIS, Fields.TOTAL_TIME, new TimeValue(time()));
+            builder.humanReadableField(Fields.SOURCE_THROTTLE_TIME_IN_MILLIS, Fields.SOURCE_THROTTLE_TIME, sourceThrottling());
+            builder.humanReadableField(Fields.TARGET_THROTTLE_TIME_IN_MILLIS, Fields.TARGET_THROTTLE_TIME, targetThrottling());
+            return builder;
+        }
+
+        @Override
+        public synchronized String toString() {
+            try {
+                XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
+                builder.startObject();
+                toXContent(builder, EMPTY_PARAMS);
+                builder.endObject();
+                return Strings.toString(builder);
+            } catch (IOException e) {
+                return "{ \"error\" : \"" + e.getMessage() + "\"}";
+            }
+        }
+        // this is the one we reported and was fixed.  This method need to have synchornixze on it
+        public File getFileDetails(String dest) {
+            return fileDetails.get(dest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **958**. At this line the following field is not protected by synchronization: `fileDetails`. Your code typically protects this field with synchronization on object: `Index.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `Index.this`. For example, like at lines: **714**, **768**, **732**, **727**, **828**, **720**, **839**, **860**, **893**, **790**, **850**, **903**, **776**, **880**, **874**, **812**, and **804**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **704**, **714**, **720**, **727**, **732**, **768**, **776**, **790**, **804**, **812**, **828**, **839**, **850**, **860**, **874**, **880**, **893**, **903**, and **932**. This is a call to `get()`. You are already protecting `get()` and `size()`. You are protecting these operations at lines: **768** and **732**.